### PR TITLE
Re-work multilocale documentation

### PR DESCRIPTION
**Pull request by mppf**
In response to some user issues with running Chapel
on an InfiniBand cluster, I've put some effort into
improving our documentation in this area. This commit:

 * improves multilocale.rst to keep the key information
   mostly on the first page
 * adds a new infiniband.rst file that shows
   how to get Chapel running on an InfiniBand cluster
 * refactors and improves documentation between launcher.rst
   and multilocale.rst.

Reviewed and tested by @ben-albrecht - thanks!

### CODE REVIEW

File: doc/release/multilocale.rst [hunk 1]
```diff
@@ -4,233 +4,196 @@
...
+
+What is GASNet?
++++++++++++++++
+
+.. _what-is-gasnet:
+
+
+GASNet is a one-sided communication and active message library being
+developed by Lawrence Berkeley National Lab and UC Berkeley.  For details,
+refer to:
```
Comment (ben-albrecht) [inline doc/release/multilocale.rst]: refer to \`GASNet's official documentation <http://gasnet.cs.berkeley.edu>`_

(we don't want to point people to [third-party/gasnet/README](https://github.com/chapel-lang/chapel/blob/master/third-party/README), which is our modified version with a bunch of licensing stuff)

**Comment by ben-albrecht** [comment]
Tested `Multilocale Chapel Execution` ssh instructions

**Comment by ben-albrecht** [comment]
I recommend a short-section in troubleshooting that mentions `scontrol show` as a way to confirm you are using the correct queue (an issue I encountered).

File: doc/release/platforms/infiniband.rst [hunk 2]
```diff
@@ -0,0 +1,178 @@
+.. _readme-infiniband:
+
+============================
+Using Chapel with InfiniBand
+============================
+
+Avoiding Slow Job Launch
+++++++++++++++++++++++++
+
+We've observed job launch hangs in some systems with InfiniBand
```
Comment (mppf) [inline doc/release/platforms/infiniband.rst]: For Tony, it takes ~ 1 minute. For other clusters, I've seen it be worse. I believe it will complete eventually, it's just users might give up...

File: doc/release/launcher.rst [hunk 3]
```diff
@@ -277,11 +277,12 @@ from SLURM. For example, the following requests 15 minutes:
...
 
 Another Slurm variable that usually needs to be set is the Slurm partition to
-use. For example, set the Slurm partition to 'debug' with the command:
+use. For example, set the Slurm partition to 'debug' with the commands:
 
 .. code-block:: bash
 
-  export SALLOC_PARTITION=chapel
+  export SALLOC_PARTITION=debug
+  export SLURM_PARTITION=$SALLOC_PARTITION
```
Comment (ben-albrecht) [inline doc/release/launcher.rst]: Curious if @gbtitus approves..

**Comment by bradcray** [comment]
When asked what I thought about revamping these docs, I said I'd defer to @gbtitus and @ronawho  who had done most of the historical maintenance of them.  If they could give this a careful post-merge review, I'd appreciate it.

**Comment by mppf** [comment]
good point, I did forget that you mentioned that.

File: doc/release/multilocale.rst [hunk 4]
```diff
@@ -4,233 +4,198 @@
...
-     portals              fast
-     other                everything
-     ===================  ====================
+  Then, consider changing several these variables:
 
+    * :ref:`CHPL_LAUNCHER <readme-launcher>` controls job launch
+    * :ref:`CHPL_COMM_SUBSTRATE <set-comm-conduit>` selects a GASNet conduit
+    * :ref:`CHPL_GASNET_SEGMENT <set-comm-segment>` indicates a memory segment
 
 #.
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: Think +# is needed here. The rendered numbering goes 0, 1, 2, 1, 2, 3, 4.

File: doc/release/multilocale.rst [hunk 5]
```diff
@@ -4,233 +4,198 @@
...
 
+     ./hello6-taskpar-dist -nl 2
 
-#. Compile your Chapel program as usual:
+   runs our Hello World program on 2 locales.
+
+What is this _real program?
++++++++++++++++++++++++++++
+
+When you compile a Chapel program for multiple locales, you should
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: Do we ever not have two binaries, for a multilocale build? The phrasing here ("should typically") indicates that sometimes we'll only have a single binary.  If that can't be, this should be rephrased.

File: doc/release/multilocale.rst [hunk 6]
```diff
@@ -4,233 +4,198 @@
...
+
+.. _set-comm-conduit:
+
+Setting CHPL_COMM_SUBSTRATE
++++++++++++++++++++++++++++
+
+Advanced users can set ``CHPL_COMM_SUBSTRATE`` to indicate the GASNet
+conduit that they wish to use. Conduits are optional components in the
+GASNet library. GASNet uses different conduits to support different
+networks. is a Novice users can leave this unset and Chapel will make a
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: spurious text: "is a" in "networks. is a Novice ..."

File: doc/release/multilocale.rst (hunk 5 above)
Comment (mppf) [inline doc/release/multilocale.rst]: Yes, if you set CHPL_LAUNCHER=none but CHPL_COMM=something you can have a multilocale build without two binaries. But that is not a typical configuration.

File: doc/release/multilocale.rst [hunk 7]
```diff
@@ -4,233 +4,198 @@
...
+psm
+    PSM for Intel's OmniPath fabric
+mpi
+    MPI - portable conduit, works on any network with MPI 1.1 or newer
+udp
+    UDP - portable conduit, works on any network with a TCP/IP stack
+    (see :ref:`Using the Portable UDP Conduit <using-udp>`)
 
-   .. code-block:: bash
+See the GASNet documentation for more information on each of these conduits.
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: Repeating the link to the GASNet website here wouldn't hurt.

File: doc/release/multilocale.rst [hunk 8]
```diff
@@ -4,233 +4,198 @@
...
+Setting CHPL_GASNET_SEGMENT
++++++++++++++++++++++++++++
 
-#. Multi-locale executions require the number of locales to be
-   specified on the command line.  Other than this, execute your
-   Chapel program as usual.  For example:
+Advanced GASNet users can set ``CHPL_GASNET_SEGMENT`` to choose a memory
+segment to use with GASNet. A GASNet segment is a region of memory that
+is expected to be used for remote memory access. The GASNet library works
+to make memory in this segment is available for accelerated memory access
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: There's a typo somewhere in "works to make memory in this segment is available", either a missing "sure" after "make", or a spurious "is".

File: doc/release/multilocale.rst [hunk 9]
```diff
@@ -4,233 +4,198 @@
...
+supported directly by network hardware.  The options are:
 
-   .. code-block:: bash
+everything
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: Say more here about the tradeoffs, in particular that "fast" means all memory is directly network referenced but program data size must be specified (or defaulted) up front, "large" means up to a specified (or default) size is directly network referenced and the rest is slower, "everything" means all network references are slower but there is no need to worry about memory size or registration.

File: doc/release/multilocale.rst [hunk 10]
```diff
@@ -4,233 +4,198 @@
...
+
+    export GASNET_ROUTE_OUTPUT=0
+
+but your mileage may vary.
+
+.. _set-comm-debugging:
+
+Advanced users may want to set ``CHPL_COMM_DEBUG`` in order to enable
+GASNet's internal sanity checking. (It is off by default.)
+Note that you'll need to re-make the compiler and runtime when changing
```
Comment (gbtitus) [inline doc/release/multilocale.rst]: Do they really have to re-make the compiler if they change CHPL_COMM_DEBUG?  Based on `git grep COMM_DEBUG` I'm not seeing that they should have to.  However, in addition to the runtime, they have to remake gasnet, which isn't automatically re-made if you re-make the runtime.

Comment (mppf) [inline doc/release/multilocale.rst]: I think it makes more sense to say "re-make Chapel" since I'm not so sure drawing a fine distinction between the compiler and runtime is actually useful in this document.

File: doc/release/launcher.rst [hunk 11]
```diff
@@ -91,122 +94,309 @@ installation-specific configuration, e.g.), you can often use the ``-v``
...
-_real binary suffix for execution
----------------------------------
+.. _using-udp:
 
-In order to support profiling tools that produce new binaries for the
-launcher to execute, the suffix of the real binary executed by the
-launcher may be changed with the ``CHPL_LAUNCHER_SUFFIX`` environment
-variable. If this variable is unset, the suffix defaults to "_real",
-matching the compiler's output.
+Using the Portable UDP Conduit
```
Comment (gbtitus) [inline doc/release/launcher.rst]: A lot of this info about the UDP conduit should be in multilocale.rst rather than here in the launcher doc.  We could talk about the amudprun and slurm-srun launchers here and reference those sections from a UDP conduit section in multilocale.rst, but much of this, for example "I'm seeing warnings from GASNet about using a higher-performance network", has nothing to do with launchers. (Plus, users probably wouldn't come here looking for it.)

Comment (mppf) [inline doc/release/launcher.rst]: Honestly I don't think it belongs in multilocale.rst OR launchers.rst. I moved most of it from multilocale.rst because I wanted the instructions there to be shorter and more overview. I'd like multilocale.rst to be a jumping-off point, depending on configuration. I really don't want it to get bogged down in the details of UDP conduit since the UDP conduit isn't even supposed to be the "main" configuration.

Put another way, I'd rather have info about Cray XC and Infiniband in multilocale.rst than info about UDP. But those things clearly don't belong there.

If we moved the inforamtion about the UDP conduit to another file, where would we put it? As a sibling to infiniband.rst in platforms/ ? Or something else?

File: doc/release/platforms/infiniband.rst [hunk 12]
```diff
@@ -0,0 +1,177 @@
+.. _readme-infiniband:
```
Comment (gbtitus) [inline doc/release/platforms/infiniband.rst]: It might be worthwhile to have a link here from the CS section in platforms/cray.

File: doc/release/platforms/infiniband.rst [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+   # Specify that ssh should be used
+   export GASNET_IBV_SPAWNER=mpi
+
+
+
+See Also
+++++++++
+
+For more information on these and other available GASNet options other
+options, including configuring to launch through MPI, please refer to
```
Comment (gbtitus) [inline doc/release/platforms/infiniband.rst]: Typo: "... other available GASNet options other options ...".

File: doc/release/multilocale.rst (hunk 10 above)
Comment (gbtitus) [inline doc/release/multilocale.rst]: Except that might not work: if you just change CHPL_COMM_DEBUG and say "make" at the top level we won't remake third-party/gasnet, because gasket's unique config path doesn't depend on the CHPL_COMM_DEBUG setting.  Maybe it ought to, but one way or another they do need to rebuild gasnet if they change this.

File: doc/release/launcher.rst (hunk 11 above)
Comment (bradcray) [inline doc/release/launcher.rst]: I could imagine overloading platforms/ to include notes on conduits as well.  Or creating a doc/conduits/ directory if we didn't like mixing our metaphors (or felt that directory was too full already).

Comment (gbtitus) [inline doc/release/launcher.rst]: Sure, definitely some aspects of the conduit and launcher choice are tightly coupled, but there are also things that are peculiar to one and don't have anything to do with the other. And there are other topics (Cray X*) where we've made a choice to push them into a subsidiary file rather than have them up in the main one, and someone else might have decided it the other way.  But we've made a conscious choice to make the top-level files as general as possible, to try to get folks up and running with a minimum of config settings. So since the UDP conduit is our portable default in so many situations I think it's best that that at least the conduit-specific aspects of using it ought to be discussed in multilocale.rst.

Comment (mppf) [inline doc/release/launcher.rst]: Greg, I'd rather delete the section "I'm seeing warnings from GASNet about using a higher-performance network" from launchers.rst and never add it anywhere than put it in multilocale.rst.

I believe everything else in "Using the Portable UDP conduit" can be viewed as being about launchers.

I'd be happy to move "Using the Portabel UDP Conduit" to its own file and link to it from both launchers.rst and multilocale.rst.

I remain unconvinced that details (other than the basic outline of instructions) of the UDP conduit belong in multilocale.rst at all. I think multilocale.rst needs to be the outline and jumping off point for multilocale execution rather than something that gives one conduit much preferential treatment.

Comment (gbtitus) [inline doc/release/launcher.rst]: Very well, I've commented on everything that mattered to me, so I think we're done now.

Comment (ben-albrecht) [inline doc/release/launcher.rst]: @bradcray  - +1 to putting UDP info into its own file. 

After some discussion offline, I propose a directory/documentation hierarchy as follows:

* Platforms
   * Cray
   * OS X
   * ...
   * Network Protocols
      * Infiniband
      * UDP

I don't think network protocols is worthy of a top-level documentation directory, aka top-level sidebar item in the [online docs](http://chapel.cray.com/docs/master/index.html)

Comment (gbtitus) [inline doc/release/launcher.rst]: Conceptually that seems okay to me.  However, the conduit concept is specific to GASNet. comm=ugni doesn't have it and our past (and potential future) comm=mpi also doesn't. The coming comm=ofi might or might not have something like it for its so-called fabric provider. So maybe instead of Infiniband and UDP conduit docs we want just one GASNet doc that talks about all the GASNet conduits, and perhaps in the future an ofi one that talks about all its fabric providers. But if the decision is to document the conduits individually I won't fight that consensus.

Comment (mppf) [inline doc/release/launcher.rst]: I think having an idea of a "network" is reasonable in that future. We could view MPI and libfabric as networks and also UGNI if we cared to. I think it makes sense to keep UGNI documented in platforms/cray for now, but I could see moving it.

But, anyway, I think part of my view here is that our documentation needs to offer tested HOWTOs for common situations. If libfabric+ some particular hardware becomes somewhat common, we'll want to document that. If libfabric becomes the better way for Chapel to work with InfiniBand (say), I'd still want users to find infiniband.rst and just see information about libfabric there.

In other words - I think it's reasonable to expect our users know what networking hardware they're trying to use. I think the exceptions are MPI and UDP - I think our users will know when they want to use MPI, and UDP will work (slowly) most everywhere. Otherwise, I'd expect they'd follow whatever suggestion we offer (ie should I use MPI, GASNet verbs, GASNet mxm, libfabric, ...? with InfiniBand?).

Comment (mppf) [inline doc/release/launcher.rst]: But, I should also say, I'd be happy to have a gasnet.rst and a libfabric.rst in such a future, to document some of the general things about those networking software libraries. If you want, we could do a gasnet.rst now. That would be nice, ecause it could make multilocale.rst even simpler...
